sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs

```diff
@@ -225,6 +226,33 @@ public class ServiceBusReceiver : IAsyncDisposable
             return messages;
         }
 
+        /// <summary>
+        /// Receives messages as an asynchronous enumerable from the entity using <see cref="ReceiveMode"/> mode.
+        /// <see cref="ReceiveMode"/> defaults to PeekLock mode. Messages will be received from the entity as
+        /// the IAsyncEnumerable is iterated. If no messages are available, this method will continue polling
+        /// until messages are available, i.e. it will never return null.
+        /// </summary>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the
+        /// request to cancel the operation.</param>
+        /// <returns>The message received.</returns>
+        public virtual async IAsyncEnumerable<ServiceBusReceivedMessage> ReceiveMessagesAsync(
+            [EnumeratorCancellation]
+            CancellationToken cancellationToken = default)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                var msg = await ReceiveMessageAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
+                if (msg == null)
+                {
+                    continue;
+                }
+                yield return msg;
+            }
+
+            // Surface the TCE to ensure deterministic behavior when cancelling.
+            throw new TaskCanceledException();
```

Comment:
It would be better to call `cancellationToken.ThrowIfCancellationRequested()` here so that user can extract canceled token from exception.